File: src/main/java/io/github/benjohnde/ankeader/transformer/AnkiTransformer.java
package io.github.benjohnde.ankeader.transformer;

import io.github.benjohnde.ankeader.anki.ApkgCard;
import io.github.benjohnde.ankeader.parser.orm.CardEntity;
import io.github.benjohnde.ankeader.parser.utils.StringUtils;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class AnkiTransformer {
    public List<ApkgCard> transformCards(List<CardEntity> cards) {
        return cards
                .stream()
                .map(card -> {
                    List<String> tags = Arrays.asList(card.getTags().trim().split(" "));
                    tags = Collections.unmodifiableList(tags);

                    List<String> slices = StringUtils.sliceBy((char) 0x1f, card.getFlds());
                    String question = slices.get(0);
                    String answer = slices.get(1);

                    return new ApkgCard(tags, answer, question);
                })
                .collect(Collectors.toList());
    }

    public void transformMedia(Map<String, String> media, String sourcePath, String destPath) {
        media.forEach((id, filename) -> {
            File outputDirectory = new File(destPath, "media");
            outputDirectory.mkdirs();

            Path src = Paths.get(sourcePath, id);
            Path dest = Paths.get(outputDirectory.getAbsolutePath(), filename);
            try {
                Files.move(src, dest);
            } catch (IOException e) {
                // file may be not present
            }
        });
    }
}

File: src/main/java/io/github/benjohnde/ankeader/transformer/AnkiTransformer.kt
package io.github.benjohnde.ankeader.transformer

import io.github.benjohnde.ankeader.anki.AnkiCard
import io.github.benjohnde.ankeader.parser.orm.CardEntity
import io.github.benjohnde.ankeader.parser.utils.StringUtils
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

class AnkiTransformer {
    fun transformCards(cards: List<CardEntity>): List<AnkiCard> {
        return cards.map {
            var tags = it.tags!!.trim().split(" ")

            if (!tags.toString().contains("Block")) {
                tags = tags.plus("Block00")
            }

            val id = it.id
            val slices = StringUtils.sliceBy(0x1f.toChar(), it.flds!!)
            val question = slices[0].trim()
            val answer = slices[1].trim()

            AnkiCard(id, tags, question, answer)
        }.filterNot {
            it.id == 1359886655736
        }
    }

    fun transformMedia(media: Map<String, String>, sourcePath: String, destPath: String) {
        media.forEach { id, filename ->
            val outputDirectory = File(destPath, "media")
            outputDirectory.mkdirs()

            val src = Paths.get(sourcePath, id)
            val dest = Paths.get(outputDirectory.absolutePath, filename)
            try {
                Files.move(src, dest)
            } catch (e: IOException) {
                // file may be not present
            }
        }
    }
}
